xnmt/decoder.py

```diff
@@ -119,13 +120,32 @@ def get_scores(self, mlp_dec_state):
     return self.vocab_projector(h_t)
 
   def calc_loss(self, mlp_dec_state, ref_action):
+    """
+        Label Smoothing is implemented with reference to Section 7 of the paper
+        "Rethinking the Inception Architecture for Computer Vision"
+        (https://arxiv.org/pdf/1512.00567.pdf)
+        """
     scores = self.get_scores(mlp_dec_state)
-    # single mode
-    if not xnmt.batcher.is_batched(ref_action):
-      return dy.pickneglogsoftmax(scores, ref_action)
-    # minibatch mode
+
+    if self.label_smoothing == 0.0:
+      # single mode
+      if not xnmt.batcher.is_batched(ref_action):
+        return dy.pickneglogsoftmax(scores, ref_action)
+      # minibatch mode
+      else:
+        return dy.pickneglogsoftmax_batch(scores, ref_action)
+
     else:
-      return dy.pickneglogsoftmax_batch(scores, ref_action)
+      log_prob = dy.log_softmax(scores)
+      if not xnmt.batcher.is_batched(ref_action):
+        pre_loss = -dy.pick(log_prob, ref_action)
+      else:
+        pre_loss = -dy.pick_batch(log_prob, ref_action)
+
+      pre_loss = dy.mean_batches(pre_loss)
```

Comment:
Why are you doing `mean_batches` here? I think you shouldn't do anything with respect to batches.

Reply:
actually, I got confused here and thought that the loss needs to be a scalar 1D value, so was doing mean_batches. I now realize that the loss needs to be present for every row.

I have updated the code now.